src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -334,38 +289,151 @@ object BlockMatrixIsDistributedMatrix extends DistributedMatrix[BlockMatrix] {
     val hadoop = m.blocks.sparkContext.hadoopConfiguration
     hadoop.mkDir(uri)
 
-    m.blocks.map { case ((i, j), m) =>
-      (new PairWriter(i, j), new MatrixWriter(m.numRows, m.numCols, m.toArray)) }
-      .saveAsSequenceFile(uri+matrixRelativePath)
-
+    writeBlocks(m, uri)
+    
     hadoop.writeDataFile(uri+metadataRelativePath) { os =>
       jackson.Serialization.write(
         BlockMatrixMetadata(m.rowsPerBlock, m.colsPerBlock, m.numRows(), m.numCols()),
         os)
     }
   }
+  
+  def writeBlocks(m: M, uri: String) {
+    val blocks = m.blocks
+    val sc = blocks.sparkContext
+    val hadoopConf = sc.hadoopConfiguration
+    
+    hadoopConf.mkDir(uri + "/blocks")
+    
+    val sHadoopConfBc = sc.broadcast(new SerializableHadoopConfiguration(hadoopConf))
+    
+    val nPartitions = blocks.getNumPartitions
+    val d = digitsNeeded(nPartitions)
+
+    val blockCount = blocks.mapPartitionsWithIndex { case (i, it) =>
+      var localBlockCount = 0L
+
+      val is = i.toString
+      assert(is.length <= d)
+      val pis = StringUtils.leftPad(is, d, "0")
+
+      sHadoopConfBc.value.value.writeDataFile(uri + "/blocks/block-" + pis) { out =>
+        it.foreach { case ((iblock, jblock), dm) => // will have one block per partition
```

Comment:
Don't comment when you can assert:

```
  block it.next()
  assert(!it.hasNext)
  ... block ...
```

Reply:
Regarding Spark iterators, we have to make sure to always call `it.hasNext` before _and_ after calls to `it.next` or things could go haywire.